### PR TITLE
[Core][Promotion] Tracking promotions usage

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Related tickets |
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Resources/config/services/promotion.xml [hunk 1]
```diff
@@ -75,5 +75,9 @@
             <argument type="service" id="sylius.factory.adjustment" />
             <argument type="service" id="sylius.integer_distributor" />
         </service>
+
+        <service id="sylius.modifier.usage_modifier" class="Sylius\Component\Core\Promotion\Modifier\OrderPromotionsUsageModifier">
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services/promotion.xml]: sylius.promotion_usage_modifier

File: src/Sylius/Component/Core/Promotion/Modifier/OrderPromotionsUsageModifier.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function increment(OrderInterface $order)
+    {
+        foreach ($order->getPromotions() as $promotion) {
+            $promotion->incrementUsed();
+        }
+
+        $this->promotionManager->flush();
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Core/Promotion/Modifier/OrderPromotionsUsageModifier.php]: I am not sure we want to flush here... The order will be flushed, so it should work without these, hm?

File: src/Sylius/Component/Core/Promotion/Modifier/OrderPromotionsUsageModifier.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function decrement(OrderInterface $order)
+    {
+        foreach ($order->getPromotions() as $promotion) {
+            $promotion->decrementUsed();
+        }
+
+        $this->promotionManager->flush();
```
Comment (pamil) [inline src/Sylius/Component/Core/Promotion/Modifier/OrderPromotionsUsageModifier.php]: Same here.

File: features/promotion/tracking_usage_limit/decreasing_promotion_usage_after_cancelling_order.feature [hunk 4]
```diff
@@ -0,0 +1,22 @@
+@managing_promotions
+Feature: Decreasing a promotion usage after cancelling an order
+    In order to properly limit promotion usage
+    As an Administrator
+    I want to have a promotion usage limit decreased after order cancellation
```
Comment (lchrusciel) [inline features/promotion/tracking_usage_limit/decreasing_promotion_usage_after_cancelling_order.feature]: This sentence is really misleading form me. `Decreased usage limit` does not equal `Decreased usage`. Limit is still same and the promotion is not used, so we should decreased usage counter, not limit.

File: features/promotion/tracking_usage_limit/increasing_promotion_usage_after_placing_order.feature [hunk 5]
```diff
@@ -0,0 +1,21 @@
+@managing_promotions
+Feature: Increasing a promotion usage after placing an order
+    In order to limit promotion usage
+    As an Administrator
+    I want to have a promotion usage limit increased after order placement
```
Comment (lchrusciel) [inline features/promotion/tracking_usage_limit/increasing_promotion_usage_after_placing_order.feature]: Same here. Word `limit` is redundant and misleading.

File: src/Sylius/Behat/Context/Setup/PromotionContext.php [hunk 6]
```diff
@@ -112,6 +112,21 @@ public function thereIsPromotion($promotionName, $promotionCode = null)
     }
 
     /**
+     * @Given there is a promotion :promotionName limited to :usageLimit usages
+     */
+    public function thereIsPromotionLimitedToUsages($promotionName, $usageLimit)
+    {
+        $promotion = $this->testPromotionFactory
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/PromotionContext.php]: I would vote to leave it in a one line

**Comment by pjedrzejewski** [comment]
Thanks Mateusz! :) Nice work!

**Pull request merged**